### PR TITLE
Adds IExecutable abstraction to core

### PR DESCRIPTION
**Pull request by PascalSenn**

### CODE REVIEW

File: src/HotChocolate/Core/src/Abstractions/IQuery.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace HotChocolate
+{
+    public interface IQuery
```
Comment (arif-hanif) [inline src/HotChocolate/Core/src/Abstractions/IQuery.cs:6]: thoughts on IDataQuery instead of IQuery so there is no confusion betweeen graphql vs data...

Comment (PascalSenn) [reply src/HotChocolate/Core/src/Abstractions/IQuery.cs]: yup that's right. We renamed it to IExecutable

**Review by michaelstaib — COMMENTED** [review]
Looks good... what are the errors?

File: src/HotChocolate/Core/src/Execution/Processing/ResolverTask.cs [hunk 2]
```diff
@@ -66,6 +66,10 @@ private async ValueTask ExecuteResolverPipelineAsync()
 
             switch (_context.Result)
             {
+                case IExecutable query:
```
Comment (arif-hanif) [inline src/HotChocolate/Core/src/Execution/Processing/ResolverTask.cs:69]: case IExecutable executable:

File: src/HotChocolate/Core/src/Execution/Processing/ResolverTask.cs [hunk 3]
```diff
@@ -66,6 +66,10 @@ private async ValueTask ExecuteResolverPipelineAsync()
...
+                    _context.Result = await query.ExecuteAsync(_context.RequestAborted);
```
Comment (arif-hanif) [inline src/HotChocolate/Core/src/Execution/Processing/ResolverTask.cs:70]: _context.Result = await executable.ExecuteAsync(_context.RequestAborted);

File: src/HotChocolate/Core/test/Types.Tests/Internal/ExtendedTypeTests.cs [hunk 4]
```diff
@@ -301,6 +301,7 @@ public void IsEqual_Object_Byte_String_False()
         [InlineData(typeof(IEnumerable<string>), "IEnumerable<String>", "String")]
         [InlineData(typeof(IReadOnlyCollection<string>), "IReadOnlyCollection<String>", "String")]
         [InlineData(typeof(IReadOnlyList<string>), "IReadOnlyList<String>", "String")]
+        [InlineData(typeof(IQuery<string>), "IQuery<String>", "String")]
```
Comment (arif-hanif) [inline src/HotChocolate/Core/test/Types.Tests/Internal/ExtendedTypeTests.cs:304]: [InlineData(typeof(IExecutable<string>), "IExecutable<String>", "String")]

File: src/HotChocolate/Core/test/Types.Tests/Internal/ExtendedTypeTests.cs [hunk 5]
```diff
@@ -328,6 +329,7 @@ public void SupportedListTypes(Type type, string listTypeName, string elementTyp
         [InlineData(typeof(CustomStringList2<string>))]
         [InlineData(typeof(ImmutableArray<string>))]
         [InlineData(typeof(IEnumerable<string>))]
+        [InlineData(typeof(IQuery<string>))]
```
Comment (arif-hanif) [inline src/HotChocolate/Core/test/Types.Tests/Internal/ExtendedTypeTests.cs:332]: [InlineData(typeof(IExecutable<string>))]

File: src/HotChocolate/Core/test/Types.Tests/Internal/ExtendedTypeTests.cs [hunk 6]
```diff
@@ -373,6 +375,20 @@ public void OptionalNullableOptionalNullableString()
             Assert.Equal("Optional<Optional<String>>!", type.ToString());
         }
 
+        [Fact]
+        public void From_IQueryScalar()
```
Comment (arif-hanif) [inline src/HotChocolate/Core/test/Types.Tests/Internal/ExtendedTypeTests.cs:379]: public void From_IExecutableScalar()

File: src/HotChocolate/Core/test/Types.Tests/Internal/ExtendedTypeTests.cs [hunk 7]
```diff
@@ -373,6 +375,20 @@ public void OptionalNullableOptionalNullableString()
...
+        {
+            // arrange
+            // act
+            ExtendedType dict = ExtendedType.FromType(
+                typeof(IQuery<string>),
```
Comment (arif-hanif) [inline src/HotChocolate/Core/test/Types.Tests/Internal/ExtendedTypeTests.cs:384]: typeof(IExecutable<string>),

File: src/HotChocolate/Core/test/Types.Tests/Internal/TypeInfoTests.cs [hunk 8]
```diff
@@ -60,6 +60,9 @@ public class TypeInfoTests
         [InlineData(typeof(IAsyncEnumerable<string>), "[String]")]
         [InlineData(typeof(IEnumerable<string>), "[String]")]
         [InlineData(typeof(IQueryable<string>), "[String]")]
+        [InlineData(typeof(IQuery<string>), "[String]")]
+        [InlineData(typeof(IQuery<int>), "[String!]")]
+        [InlineData(typeof(IQuery<int?>), "[String]")]
```
Comment (arif-hanif) [inline src/HotChocolate/Core/test/Types.Tests/Internal/TypeInfoTests.cs:65]: IExectuable

File: src/HotChocolate/Core/test/Types.Tests/Internal/TypeInfoTests.cs [hunk 9]
```diff
@@ -84,6 +87,7 @@ public class TypeInfoTests
         [InlineData(typeof(IEnumerable<string>), "[String]")]
         [InlineData(typeof(IReadOnlyCollection<string>), "[String]")]
         [InlineData(typeof(IReadOnlyList<string>), "[String]")]
+        [InlineData(typeof(IQuery<string>), "[String]")]
```
Comment (arif-hanif) [inline src/HotChocolate/Core/test/Types.Tests/Internal/TypeInfoTests.cs:90]: IExecutable

File: src/HotChocolate/Core/test/Types.Tests/Internal/TypeInfoTests.cs [hunk 10]
```diff
@@ -538,6 +597,14 @@ public class Nullability
 
             public ICollection<string>? NullableCollectionNonNullElement() => default;
 
+            public IQuery<string> NonNullQueryNonNullElement() => default!;
+
+            public IQuery<string?> NonNullQueryNullableElement() => default!;
+
+            public IQuery<string?>? NullableQueryNullableElement() => default;
+
+            public IQuery<string>? NullableQueryNonNullElement() => default;
```
Comment (arif-hanif) [inline src/HotChocolate/Core/test/Types.Tests/Internal/TypeInfoTests.cs:606]: IExecutable

File: src/HotChocolate/Core/test/Types.Tests/Internal/ExtendedTypeTests.cs (hunk 4 above)
Comment (PascalSenn) [reply src/HotChocolate/Core/test/Types.Tests/Internal/ExtendedTypeTests.cs:304]: that is something we cannot allow
we need to know the type for the list :)

**Comment by PascalSenn** [comment]
@arif-hanif  Thanks for the review!

**Review by arif-hanif — APPROVED** [review]
approved

**Pull request merged**